# Traffic Sign Recognition

## What is Traffic Signs Recognition?
There are several different types of traffic signs like speed limits, no entry, traffic signals, turn left or right, children crossing, no passing of heavy vehicles, etc. Traffic signs classification is the process of identifying which class a traffic sign belongs to.


## Aim of the project
We have to build a Deep Neural Network model that can classify traffic signs present in the image into different categories. With this model, we should be able to read and understand traffic signs which are a very important task for all autonomous vehicles.

Importing all necessary libraries

Importing data and preprocessing

Converting image and label lists into numpy arrays

Splitting data and labels into training and testing dataset

CNN Model Building

Compilation of the model

fitting of the model

## Importing all necessary libraries

In [1]:
from tensorflow.keras.layers import Conv2D, MaxPool2D, Dense, Flatten, Dropout
from tensorflow.keras.models import Sequential, load_model
from sklearn.model_selection import train_test_split
from tensorflow.keras.utils import to_categorical
import matplotlib.pyplot as plt
import tensorflow as tf
from PIL import Image
import pandas as pd 
import numpy as np 
import cv2
import os

## Importing data and preprocessing

### Info on data 
The dataset contains more than 50,000 images of different traffic signs.

It is further classified into 43 different classes.

The ‘train’ folder contains 43 folders each representing a different class. The range of the folder is from 0 to 42. 


In [4]:
data = []
labels = []
categories = 43
height = 30
width = 30
curr_path = os.getcwd()                 # The method os. getcwd() in Python returns the current working directory of a process. 

for i in range(categories):
    path = os.path.join(curr_path,'Dataset/train',str(i))       #joins one or more path components
    images = os.listdir(path)                                   #lists files and directories in the given path 
    for a in images:    
        try:
            image = Image.open(path + '//'+ a)                  #opening other image file     
            image = image.resize((height,width))                #resizing the image to maintain uniform
            image = np.array(image)                             #getting array of images using numpy
            data.append(image)                                  #appending images data to data list
            labels.append(i)                                    #appending the labels list
        except:
            print("Error loading image")

## Converting image and label lists into numpy arrays

We need to convert the data and label list into numpy arrays for feeding to the model.

In [5]:
data = np.array(data)                          #data list into numpy array
labels = np.array(labels)                      #label list into numpy array
print(data.shape, labels.shape)                #shape of data and label

(39209, 30, 30, 3) (39209,)



The shape of data is (39209, 30, 30, 3)

which means that there are 39209 images 

image size 30×30 pixels 

The last 3 defines the data contains colored images i,e (RGB value).

## Splitting data and labels into training and testing dataset

In [7]:
X_train, X_test, y_train, y_test = train_test_split(data,          #train_test_split() method to split training and testing data
                                                    labels, 
                                                    test_size=0.2, 
                                                    random_state=30)

print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)    # shape of training and testing 

(31367, 30, 30, 3) (7842, 30, 30, 3) (31367,) (7842,)


test size is 0.2 i,e 20 % data
### For training
The shape of data is (31367, 30, 30, 3)

which means that there are 31367 images

image size 30×30 pixels

The last 3 defines the data contains colored images i,e (RGB value).

### for testing 
The shape of data is (7842, 30, 30, 3)

which means that there are 7842 images

image size 30×30 pixels

The last 3 defines the data contains colored images i,e (RGB value).

## Converting the labels
to_categorical Converts a class vector (integers) to binary class matrix

In [10]:
y_train = to_categorical(y_train, 43)
y_test = to_categorical(y_test, 43)

# CNN Model Building

A Sequential model is appropriate for a plain stack of layers where each layer has exactly one input tensor and one output tensor.

conv2d() is the TensorFlow function you can use to build a 2D convolutional layer as part of your CNN architecture.

ReLU for short is a piecewise linear function that will output the input directly if it is positive, otherwise, it will output zero.

Max pooling is a type of operation that is typically added to CNNs following individual convolutional layers. When added to a model, max pooling reduces the dimensionality of images by reducing the number of pixels in the output from the previous convolutional layer.

Dropout is a technique used to prevent a model from overfitting. Dropout works by randomly setting the outgoing edges of hidden units (neurons that make up hidden layers) to 0 at each update of the training phase.

Dense layer is a fully connected layer, meaning all the neurons in a layer are connected to those in the next layer.

The softmax function is used as the activation function in the output layer of neural network models that predict a multinomial probability distribution. That is, softmax is used as the activation function for multi-class classification problems.

In [11]:
model = Sequential()

model.add(Conv2D(filters=32, kernel_size=(5,5), 
                 activation='relu', 
                 input_shape=X_train.shape[1:]))

model.add(Conv2D(filters=32, kernel_size=(5,5),
                 activation='relu'))

model.add(MaxPool2D(pool_size=(2, 2)))
model.add(Dropout(rate=0.25))

model.add(Conv2D(filters=32, kernel_size=(3, 3), 
                 activation='relu'))

model.add(Conv2D(filters=32, kernel_size=(3, 3), 
                 activation='relu'))

model.add(MaxPool2D(pool_size=(2, 2)))
model.add(Dropout(rate=0.3))

model.add(Flatten())                                          

model.add(Dense(190, activation='relu')) 

model.add(Dropout(rate=0.5))

model.add(Dense(43, activation='softmax'))

# Compilation of the model


### Compile defines the loss function, the optimizer and the metrics

In [12]:
model.compile(loss='categorical_crossentropy', 
              optimizer='adam', 
              metrics=['accuracy'])

## Fitting of a model

### Trains the model for a fixed number of epochs

In [13]:
model.fit(X_train, y_train, batch_size=32, 
          epochs=15, 
          validation_data=(X_test, y_test))

Epoch 1/15
981/981 [==============================] - 94s 95ms/step - loss: 3.7343 - accuracy: 0.2080 - val_loss: 1.4752 - val_accuracy: 0.5925
Epoch 2/15
981/981 [==============================] - 89s 91ms/step - loss: 1.5402 - accuracy: 0.5169 - val_loss: 0.7089 - val_accuracy: 0.8045
Epoch 3/15
981/981 [==============================] - 90s 92ms/step - loss: 0.9317 - accuracy: 0.7135 - val_loss: 0.2731 - val_accuracy: 0.9227
Epoch 4/15
981/981 [==============================] - 90s 92ms/step - loss: 0.6147 - accuracy: 0.8157 - val_loss: 0.2000 - val_accuracy: 0.9454
Epoch 5/15
981/981 [==============================] - 91s 92ms/step - loss: 0.4929 - accuracy: 0.8503 - val_loss: 0.1549 - val_accuracy: 0.9601
Epoch 6/15
981/981 [==============================] - 95s 97ms/step - loss: 0.4246 - accuracy: 0.8739 - val_loss: 0.1807 - val_accuracy: 0.9526
Epoch 7/15
981/981 [==============================] - 91s 92ms/step - loss: 0.4085 - accuracy: 0.8793 - val_loss: 0.1050 - val_accuracy:

## Our model got a 93% accuracy on the training dataset

### Model Evaluate
Evaluation is a process during development of the model to check whether the model is best fit for the given problem and corresponding data

In [15]:
model.evaluate(X_test, y_test, verbose=2)

246/246 - 4s - loss: 0.0663 - accuracy: 0.9824


[0.0663302093744278, 0.9824024438858032]

### Probability Model

In [21]:
probability_model = tf.keras.Sequential([model,
                                        tf.keras.layers.Softmax()])

In [23]:
probability_model(X_test[:4])

<tf.Tensor: shape=(4, 43), dtype=float32, numpy=
array([[0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.06078678, 0.02236221,
        0.02236221, 0.02236221, 0.02236221],
       [0.02259477, 0.02259578, 0.0226051 , 0.0229452 , 0.02259478,
        0.02286181, 0.0225948 , 0.02259884, 0.02284053, 0.02720262,
        0.02265869, 0.02259479, 0.02275968, 0.02370743, 0.02259481,
        0.02277599, 0.0226652 , 0.02259478, 0.02259479, 0.02259477,
        0.02259581, 0.02259477, 0.02259477, 0.02259478, 0.02259477,
        0.02259597, 0.

### Prediction Model

In [26]:
predictions = probability_model.predict(X_test)

In [28]:
predictions[:4]

array([[0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.02236221, 0.02236221,
        0.02236221, 0.02236221, 0.02236221, 0.06078678, 0.02236221,
        0.02236221, 0.02236221, 0.02236221],
       [0.02259478, 0.02259578, 0.0226051 , 0.0229452 , 0.02259478,
        0.02286181, 0.0225948 , 0.02259884, 0.02284053, 0.02720262,
        0.02265868, 0.02259478, 0.02275967, 0.02370743, 0.02259481,
        0.022776  , 0.0226652 , 0.02259478, 0.02259479, 0.02259478,
        0.02259581, 0.02259478, 0.02259478, 0.02259478, 0.02259478,
        0.02259597, 0.02259478, 0.02259478, 0.02259478, 0.02259479,
   

### Model Saving 

In [32]:
model.save("TSR")

INFO:tensorflow:Assets written to: TSR\assets
